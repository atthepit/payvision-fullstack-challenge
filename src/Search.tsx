import React, { FormEvent } from "react";
import { Filters } from "./transactions/transactions";

interface Props {
  onSearch: (filters: Filters) => void;
}

class Search extends React.Component<Props> {
  public handleSearchSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const form = event.currentTarget;
    const filters: Filters = {};

    const action = form.elements.namedItem("action");
    if (action) {
      const value = (action as HTMLSelectElement).value;
      if (value) {
        filters.action = value;
      }
    }

    const currencyCode = form.elements.namedItem("currencyCode");
    if (currencyCode) {
      const value = (currencyCode as HTMLSelectElement).value;
      if (value) {
        filters.currencyCode = value;
      }
    }

    this.props.onSearch(filters);
  };

  public render() {
    return (
      <form onSubmit={this.handleSearchSubmit}>
        <label htmlFor="action">Action</label>
        <select name="action" id="action">
          <option value="">Transaction type</option>
          <option value="payment">Payment</option>
          <option value="credit">Credit</option>
        </select>

        <label htmlFor="currencyCode">Currency Code</label>
        <select name="currencyCode" id="currencyCode">
          <option value="">Currency</option>
          <option>EUR</option>
          <option>USD</option>
          <option>JPY</option>
        </select>

        <button type="submit">Search</button>
      </form>
    );
  }
}

export default Search;
