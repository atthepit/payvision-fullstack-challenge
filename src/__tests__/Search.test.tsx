import React from "react";
import { render, fireEvent, cleanup } from "react-testing-library";
import Search from "../Search";

afterEach(cleanup);

test("call onSearch with filters on submit", () => {
  const onSearch = jest.fn();
  const { container, getByLabelText, getByText } = render(
    <Search onSearch={onSearch} />
  );

  const form = container.querySelector("form");
  if (form == null) {
    fail("Expected form to be rendered ");
    return;
  }

  const actionNode = getByLabelText("Action") as HTMLSelectElement;
  const currencyCodeNode = getByLabelText("Currency Code") as HTMLSelectElement;
  const searchButton = getByText("Search");

  actionNode.value = "payment";
  currencyCodeNode.value = "EUR";
  fireEvent.click(searchButton);

  expect(onSearch).toHaveBeenCalledTimes(1);
  expect(onSearch).toHaveBeenCalledWith({
    action: actionNode.value,
    currencyCode: currencyCodeNode.value
  });
});

test("call onSearch with empty object if no filter is selected", () => {
  const onSearch = jest.fn();
  const { container } = render(<Search onSearch={onSearch} />);

  const form = container.querySelector("form");
  if (form == null) {
    fail("Expected form to be rendered ");
    return;
  }

  fireEvent.submit(form);

  expect(onSearch).toHaveBeenCalledTimes(1);
  expect(onSearch).toHaveBeenCalledWith({});
});

test("call onSearch with one filter if only one filter is selected", () => {
  const onSearch = jest.fn();
  const { container, getByLabelText, getByText } = render(
    <Search onSearch={onSearch} />
  );

  const form = container.querySelector("form");
  if (form == null) {
    fail("Expected form to be rendered ");
    return;
  }

  const actionNode = getByLabelText("Action") as HTMLSelectElement;

  actionNode.value = "payment";
  fireEvent.submit(form);

  expect(onSearch).toHaveBeenCalledTimes(1);
  expect(onSearch).toHaveBeenCalledWith({
    action: actionNode.value
  });
});
