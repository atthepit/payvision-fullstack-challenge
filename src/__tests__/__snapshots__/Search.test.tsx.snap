// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  margin-bottom: 1rem;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.emotion-5 > * {
  margin-left: 2%;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.emotion-1 {
  border: 1px solid #213d8f;
  color: #213d8f;
  font-weight: 600;
  padding: 0.4rem;
}

.emotion-4 {
  background-color: #8ec03f;
  border: none;
  color: #fff;
  padding: 0.5rem 1rem;
}

.emotion-4:hover {
  cursor: pointer;
}

.emotion-4:hover,
.emotion-4:focus {
  background-color: hsl( 83, 51%, 55% );
}

<form
  class="emotion-5"
>
  <label
    class="emotion-0"
    for="action"
  >
    Action
  </label>
  <select
    class="emotion-1"
    id="action"
    name="action"
  >
    <option
      value=""
    >
      Transaction type
    </option>
    <option
      value="payment"
    >
      Payment
    </option>
    <option
      value="credit"
    >
      Credit
    </option>
  </select>
  <label
    class="emotion-0"
    for="currencyCode"
  >
    Currency Code
  </label>
  <select
    class="emotion-1"
    id="currencyCode"
    name="currencyCode"
  >
    <option
      value=""
    >
      Currency
    </option>
    <option>
      EUR
    </option>
    <option>
      USD
    </option>
    <option>
      JPY
    </option>
  </select>
  <button
    class="emotion-4"
    type="submit"
  >
    Search
  </button>
</form>
`;
